volatility3/framework/automagic/mac.py

```diff
@@ -168,80 +168,223 @@ def find_aslr(
-                context, layer_name, progress_callback
-            )
-        else:
-            offset_generator = [(compare_banner_offset, compare_banner)]
+        """
+        References :
+         - https://github.com/apple-open-source/macos/blob/14.3/xnu/osfmk/x86_64/lowmem_vectors.c#L78
+         - https://github.com/apple-open-source/macos/blob/14.3/xnu/osfmk/x86_64/lowglobals.h#L50
+        """
+
+        symbol_table_object: mac.MacKernelIntermedSymbols = context.symbol_space[
+            symbol_table
+        ]
+        lowglo_json_address = symbol_table_object.get_symbol("lowGlo").address
+        lowglo_phys_offset = cls.virtual_to_physical_address(lowglo_json_address)
 
         aslr_shift = 0
+        for offset in cls._lowglo_scan_generator(
+            context, layer_name, progress_callback
+        ):
 
-        for offset, banner in offset_generator:
-            banner_major, banner_minor = [int(x) for x in banner[22:].split(b".")[0:2]]
+            aslr_shift_candidate = offset - lowglo_phys_offset
+            if aslr_shift_candidate & 0xFFF != 0:
+                continue
 
-            tmp_aslr_shift = offset - cls.virtual_to_physical_address(
-                version_json_address
+            # We can use the convenient module methods, with virtual addresses in physical space,
+            # because the shift is constant and known.
+            module = context.module(
+                symbol_table,
+                layer_name,
+                aslr_shift_candidate - cls._KERNEL_MIN_ADDRESS,
             )
 
-            major_string = context.layers[layer_name].read(
-                version_major_phys_offset + tmp_aslr_shift, 4
-            )
-            major = struct.unpack("<I", major_string)[0]
+            # https://github.com/apple-open-source/macos/blob/14.3/xnu/osfmk/i386/i386_vm_init.c#L282
+            vm_kernel_slide_candidate = module.object_from_symbol("vm_kernel_slide")
+            mh_fileset_kernel_cache_check = False
+            try:
+                mh_fileset_kernel_cache_check = cls.detect_mh_fileset_kernel_cache(
+                    context,
+                    symbol_table,
+                    layer_name,
+                    aslr_shift_candidate,
+                )
+            except exceptions.SymbolError as e:
+                vollog.log(constants.LOGLEVEL_VVVV, e)
+            except exceptions.InvalidAddressException:
+                continue
 
-            if major != banner_major:
+            if mh_fileset_kernel_cache_check:
+                vm_kernel_slide_candidate = (
+                    cls.vm_kernel_slide_kernel_cache_calculations(
+                        context,
+                        symbol_table,
+                        layer_name,
+                        aslr_shift_candidate,
+                        vm_kernel_slide_candidate,
+                    )
+                )
+
+            if vm_kernel_slide_candidate & 0xFFF != 0:
```

Comment:
The page size might be architecture independant, even if it's 4kB for macOS x86/64, in the future it might be 16kB or 64kB for AArch64. To determine the page size, we would typically need to read a symbol from the kernel (if it even exists), but is a bit of a "snake biting its tail" situation.

This logic would need to be updated at two other places in this file (dtb aligned and aslr shift aligned).

Reply:
Well, given it can be different we should at least parameterize these.  The page size should be dependent on the architecture, so a value taken from the layer (which is why the layer has the page_size value)...

Reply:
The `LimeLayer`, `VMWareLayer` , `FileLayer` etc. is embedding the page size ? I don't know how this could be relevant to higher layers (Intel, AArch64) ?

But at least parameterizing it would be more uniform, indeed !